chrome/content/zotero/xpcom/attachments.js

```diff
@@ -2241,95 +2241,195 @@ Zotero.Attachments = new function(){
 	 * (Optional) |formatString| specifies the format string -- otherwise
 	 * the 'attachmentRenameFormatString' pref is used
 	 *
-	 * Valid substitution markers:
-	 *
-	 * %c -- firstCreator
-	 * %y -- year (extracted from Date field)
-	 * %t -- title
-	 *
-	 * Fields can be truncated to a certain length by appending an integer
-	 * within curly brackets -- e.g. %t{50} truncates the title to 50 characters
-	 *
 	 * @param {Zotero.Item} item
 	 * @param {String} formatString
 	 */
 	this.getFileBaseNameFromItem = function (item, formatString) {
 		if (!(item instanceof Zotero.Item)) {
 			throw new Error("'item' must be a Zotero.Item");
 		}
-		
+
 		if (!formatString) {
 			formatString = Zotero.Prefs.get('attachmentRenameFormatString');
 		}
-		
-		// Replaces the substitution marker with the field value,
-		// truncating based on the {[0-9]+} modifier if applicable
-		function rpl(field, str) {
-			if (!str) {
-				str = formatString;
-			}
-			
-			switch (field) {
-				case 'creator':
-					field = 'firstCreator';
-					var rpl = '%c';
+
+		const getSlicedCreatorsOfType = (creatorType, slice) => {
+			let creatorTypeIDs;
+			switch (creatorType) {
+				case 'author':
+				case 'authors':
+					creatorTypeIDs = [Zotero.CreatorTypes.getPrimaryIDForType(item.itemTypeID)];
 					break;
-					
-				case 'year':
-					var rpl = '%y';
+				case 'editor':
+				case 'editors':
+					creatorTypeIDs = [Zotero.CreatorTypes.getID('editor'), Zotero.CreatorTypes.getID('seriesEditor')];
 					break;
-					
-				case 'title':
-					var rpl = '%t';
+				default:
+				case 'creator':
+				case 'creators':
+					creatorTypeIDs = null;
 					break;
 			}
 			
-			var value;
-			switch (field) {
+			if (slice === 0) {
+				return [];
+			}
+			const matchingCreators = creatorTypeIDs === null
+				? item.getCreators()
+				: item.getCreators().filter(c => creatorTypeIDs.includes(c.creatorTypeID));
+			const slicedCreators = slice > 0
+				? matchingCreators.slice(0, slice)
+				: matchingCreators.slice(slice);
+
+			if (slice < 0) {
+				slicedCreators.reverse();
+			}
+			return slicedCreators;
+		};
+
+
+		const common = (value, { truncate = false, prefix = '', suffix = '', case: textCase = '' } = {}) => {
+			if (value === '' || value === null || typeof value === 'undefined') {
+				return '';
+			}
+			if (truncate) {
+				value = value.substr(0, truncate);
+			}
+			if (prefix) {
+				value = prefix + value;
+			}
+			if (suffix) {
+				value += suffix;
+			}
+			switch (textCase) {
+				case 'upper':
+					value = value.toUpperCase();
+					break;
+				case 'lower':
+					value = value.toLowerCase();
+					break;
+				case 'sentence':
+					value = value.slice(0, 1).toUpperCase() + value.slice(1);
+					break;
 				case 'title':
-					value = item.getField('title', false, true);
+					value = value.split(' ').map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase()).join(' ');
```

Comment:
Better to use `Zotero.Utilities.capitalizeTitle(str, true)`, which has some (English) skip words and also will ignore words that already contain capitalization. (The `true` for `force` is necessary until we remove the pointless `capitalizeTitles` pref.)